### ISSUE TITLE
Spectral Normalization

### ISSUE THREAD

**Issue by MoustafaMeshry** [tensorflow/tensorflow#41216]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): TF 2.2
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Need to support Spectral Normalization as a tf.keras.layers.Wrapper to wrap dense and convolution layers.

**Will this change the current api? How?**
No

**Who will benefit with this feature?**
Spectral Normalization (https://arxiv.org/abs/1802.05957) is widely used for training generative adversarial networks. Many recent research use it and it is supported in pytorch ([torch.nn.utils.spectral_norm](https://pytorch.org/docs/master/generated/torch.nn.utils.spectral_norm.html)) for a long time now, but still no supported in tensorflow.

**Any Other info.**
The main difficulty in implementing Spectral Normalization is that it needs to update the layer kernel (e.g. the kernel of tf.keras.layers.Conv2D) using tf.Variable.assign() which is not permitted within a Replica context (e.g. building a convolution layer under tf.distribute.MirroredStrategy).

**Comment by rxng8** [comment]
can we just treat spectral normalization as one layer in the graph?

**Comment by MoustafaMeshry** [comment]
> can we just treat spectral normalization as one layer in the graph? so we don't need to worry about modifying the previous layers

This is what I've done for now. I've extended tf.keras.layers.Conv2D and tf.keras.layers.Dense directly. This way I need to store the value of a single variable used for normalization, and this variable is created with `aggregation=tf.VariableAggregation.ONLY_FIRST_REPLICA`, since its value doesn't depend on different minibatches to different replicas.

**Comment by pjpratik** [comment]
@MoustafaMeshry Can you please check [this](https://www.tensorflow.org/addons/api_docs/python/tfa/layers/SpectralNormalization) link for Spectral Normalization available in tensorflow addons. Please let us know if this help. Thank you.
